src/ripple/unl/tests/qalloc.h

```diff
@@ -0,0 +1,396 @@
+    auto pad = [](void const* p, std::size_t align)
+    {
+        auto const i = reinterpret_cast<
+            std::uintptr_t>(p);
+        return (align - (i % align)) % align;
+    };
+
+    auto const n0 =
+        pad(free_ + sizeof(block*), align);
+    auto const n1 =
+        n0 + sizeof(block*) + bytes;
+    if (remain_ < n1)
+        return nullptr;
+    auto p = reinterpret_cast<block**>(
+        free_ + n0 + sizeof(block*));
+    assert(pad(p - 1,
+        std::alignment_of<block*>::value) == 0);
+    p[-1] = this;
+    ++count_;
+    free_ += n1;
+    remain_ -= n1;
+    return p;
+}
+
+template <class _>
+bool
+qalloc_impl<_>::block::deallocate()
+{
+    --count_;
+    if (count_ > 0)
+        return false;
+    remain_ = bytes_;
+    free_ = reinterpret_cast<
+        std::uint8_t*>(this + 1);
+    return true;
+}
+
+template <class _>
+template <class U>
+inline
+auto
+qalloc_impl<_>::block::stash (U const* u) ->
+    block*&
+{
+    static auto const a =
+        std::alignment_of<block*>::value;
+    return *reinterpret_cast<block**>(
+        a * ((reinterpret_cast<std::uintptr_t>(u) - sizeof(block*)) / a));
+}
+
+template <class _>
+qalloc_impl<_>::~qalloc_impl()
+{
+    if (used_)
+        std::free(used_);
+    while (free_)
+    {
+        auto const next = free_->next;
+        free_->~block();
+        std::free(free_);
+        free_ = next;
+    }
+}
+
+template <class _>
+void*
+qalloc_impl<_>::allocate(
+    std::size_t bytes, std::size_t align)
+{
+    if (used_)
+    {
+        auto const p =
+            used_->allocate(bytes, align);
+        if (p)
+            return p;
+        used_ = nullptr;
+    }
+    if (free_)
+    {
+        auto const p =
+            free_->allocate(bytes, align);
+        if (p)
+        {
+            used_ = free_;
+            free_ = free_->next;
+            return p;
+        }
+    }
+    // VFALCO Formula needs audit
+    auto const n = std::max<std::size_t>(
+        block_size, sizeof(block) +
+            sizeof(block*) + bytes + 2 * align);
+    block* const b =
+        new(std::malloc(n)) block(n);
+    if (! b)
+        throw std::bad_alloc();
+    used_ = b;
+    // VFALCO This has to succeed
+    return used_->allocate(bytes, align);
+}
+
+template <class _>
+void
+qalloc_impl<_>::deallocate (void* p)
+{
+    auto const block = block::stash(p);
+    if (block->deallocate())
+    {
+        if (used_ == block)
+            used_ = nullptr;
+        block->next = free_;
+        free_ = block;
+    }
+}
+
+} // detail
+
+//------------------------------------------------------------------------------
+
+template <class T>
+qalloc_type<T>::qalloc_type()
+    : impl_ (std::make_shared<
+        detail::qalloc_impl<>>())
+{
+}
+
+template <class T>
+template <class U>
+qalloc_type<T>::qalloc_type(
+        qalloc_type<U> const& u)
+    : impl_ (u.impl_)
+{
+}
+
+template <class T>
+T*
+qalloc_type<T>::allocate (std::size_t n)
+{
+    if (n > std::numeric_limits<
```

Comment:
Nit: almost identical to the code below. Could we just `return alloc<T>(n)` or even a single `allocate` function with `template <class U=T>`.